rustls/src/client/client_conn.rs

```diff
@@ -255,6 +262,24 @@ impl ClientConfig {
         }
     }
 
+    /// Enable Encrypted Client Hello (ECH) support with the given config.
+    ///
+    /// If the client configuration has enabled TLS 1.2, this function will return an error. ECH
+    /// may only be used with TLS 1.3+.
+    pub fn enable_ech(&mut self, config: EchConfig) -> Result<(), Error> {
```

Comment:
So this is a bit odd in the sense that so far, we've built `ClientConfig`s by either going through the `ConfigBuilder` or just setting `pub` fields, and this is doing neither. I was thinking about whether it's possible to make the 1.2 xor ECH choice a compile-time failure but that probably involves adding another config builder state which we may not want?

Maybe we could provide a `with_ech()` method in in the `ConfigBuilder<S, WantsVersions>` state, which forces 1.3 only?

Reply:
I've tried a `with_ech` approach. I think it's better but there's a couple rough spots left:

1. We don't want to define this on `ConfigBuilder<S, WantsVersions>`, but `ConfigBuilder<ClientConfig, WantsVersions>` - the server and client configuration for ECH are quite different (e.g. a server might have multiple lists of ECH config, and will want to select a subset for offering as retry configs to facilitate graceful turn-down of old configs).

2. As a result, we need to stash a bit of client-specific state from the `impl ConfigBuilder<ClientConfig, WantsVersions> { fn with_ech(...) ... }` fn. That's not something supported by the builder architecture in general, only the last step of the build process can put things into the `ClientConfig`, and the `WantsVersion` state where we normally stash in-progress-build state is shared for both the client/server. The end result is the state gets an `Option` field that's specific to clients and ignored by builders. A bit gross, but I think workable since it's an internal detail.

3. There's _still_ an incompatible config condition we can't detect until we try to actually do ECH: the `ClientConfig` struct has a `enable_sni` `bool` field that could be set to `false` _after_ the config is built with ECH support. Configuring ECH but not sending SNI doesn't make sense, and this manifests as a new `EncryptedClientHelloError::SniRequired` error being returned from `hs.rs`'s `start_handshake` when ECH was configured but SNI disabled.

I don't think the above are deal-breakers, just wanted to call out that there's still some room for improvement.